File: go/internal/app/baseline.go
package app

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/dutifuldev/slophammer/go/internal/rules"
)

const baselineFileName = "slophammer-baseline.json"

// BaselineMode selects how check treats the checked-in baseline file.
type BaselineMode int

const (
	BaselineOff BaselineMode = iota
	BaselineCheck
	BaselineWrite
)

type baselineEntry struct {
	RuleID string `json:"rule_id"`
	Path   string `json:"path"`
}

type baselineFile struct {
	Version  int             `json:"version"`
	Findings []baselineEntry `json:"findings"`
}

// applyBaselineCheck applies a checked-in baseline to a report: matched
// findings are marked baselined and stop affecting OK; stale entries are an
// error so the ratchet can only shrink. Matching is on rule_id plus path,
// never message.
func applyBaselineCheck(root string, report *rules.Report) error {
	baseline, err := readBaselineFile(root)
	if err != nil {
		return err
	}
	unmatched := entryDifference(baseline, nil)
	ok := true
	for i := range report.Findings {
		entry := entryOf(report.Findings[i])
		if _, matched := baseline[entry]; matched {
			report.Findings[i].Baselined = true
			delete(unmatched, entry)
		} else {
			ok = false
		}
	}
	if len(unmatched) > 0 {
		return fmt.Errorf("baseline contains resolved findings; rewrite it: %s", joinedEntries(unmatched))
	}
	report.OK = ok
	return nil
}

// writeBaselineFile records current findings as the baseline. It refuses to
// write a superset of an existing baseline and reports the added and removed
// entries, so debt is recorded once, reviewed, and only ever reduced.
func writeBaselineFile(root string, report rules.Report) (string, error) {
	current := map[baselineEntry]struct{}{}
	for _, finding := range report.Findings {
		current[entryOf(finding)] = struct{}{}
	}
	previous, err := readBaselineFile(root)
	if err != nil {
		previous = nil
	}
	added := entryDifference(current, previous)
	removed := entryDifference(previous, current)
	if len(previous) > 0 && len(added) > 0 && len(removed) == 0 {
		return "", fmt.Errorf("baseline write would grow the baseline; fix the new findings instead: %s", joinedEntries(added))
	}
	if err := writeBaselineEntries(root, current); err != nil {
		return "", err
	}
	return baselineWriteSummary(len(current), added, removed), nil
}

func writeBaselineEntries(root string, entries map[baselineEntry]struct{}) error {
	serialized, err := json.MarshalIndent(baselineFile{Version: 1, Findings: sortedBaselineEntries(entries)}, "", "  ")
	if err != nil {
		return fmt.Errorf("baseline write failed: %w", err)
	}
	if err := os.WriteFile(filepath.Join(root, baselineFileName), append(serialized, '\n'), 0o600); err != nil {
		return fmt.Errorf("baseline write failed: %w", err)
	}
	return nil
}

func baselineDebtLine(report rules.Report) string {
	baselined := 0
	for _, finding := range report.Findings {
		if finding.Baselined {
			baselined++
		}
	}
	return fmt.Sprintf("%d findings baselined; %d new\n", baselined, len(report.Findings)-baselined)
}

func readBaselineFile(root string) (map[baselineEntry]struct{}, error) {
	// #nosec G304 -- the baseline lives at a fixed name under the check root.
	content, err := os.ReadFile(filepath.Join(root, baselineFileName))
	if err != nil {
		return nil, errors.New("baseline file " + baselineFileName + " is missing")
	}
	decoder := json.NewDecoder(bytes.NewReader(content))
	decoder.DisallowUnknownFields()
	var file baselineFile
	if err := decoder.Decode(&file); err != nil {
		return nil, fmt.Errorf("baseline parse failed: %w", err)
	}
	if file.Version != 1 {
		return nil, errors.New("baseline version must be 1")
	}
	entries := map[baselineEntry]struct{}{}
	for _, entry := range file.Findings {
		entries[entry] = struct{}{}
	}
	return entries, nil
}

func entryOf(finding rules.Finding) baselineEntry {
	return baselineEntry{RuleID: finding.RuleID, Path: finding.Path}
}

func entryDifference(left map[baselineEntry]struct{}, right map[baselineEntry]struct{}) map[baselineEntry]struct{} {
	diff := map[baselineEntry]struct{}{}
	for entry := range left {
		if _, ok := right[entry]; !ok {
			diff[entry] = struct{}{}
		}
	}
	return diff
}

func sortedBaselineEntries(entries map[baselineEntry]struct{}) []baselineEntry {
	sorted := make([]baselineEntry, 0, len(entries))
	for entry := range entries {
		sorted = append(sorted, entry)
	}
	sort.Slice(sorted, func(i, j int) bool {
		if sorted[i].RuleID == sorted[j].RuleID {
			return sorted[i].Path < sorted[j].Path
		}
		return sorted[i].RuleID < sorted[j].RuleID
	})
	return sorted
}

func joinedEntries(entries map[baselineEntry]struct{}) string {
	joined := make([]string, 0, len(entries))
	for _, entry := range sortedBaselineEntries(entries) {
		joined = append(joined, entry.RuleID+" at "+entry.Path)
	}
	return strings.Join(joined, ", ")
}

func baselineWriteSummary(total int, added map[baselineEntry]struct{}, removed map[baselineEntry]struct{}) string {
	var summary strings.Builder
	_, _ = fmt.Fprintf(&summary, "baseline written: %d finding(s)\n", total)
	for _, entry := range sortedBaselineEntries(added) {
		_, _ = fmt.Fprintf(&summary, "added: %s at %s\n", entry.RuleID, entry.Path)
	}
	for _, entry := range sortedBaselineEntries(removed) {
		_, _ = fmt.Fprintf(&summary, "removed: %s at %s\n", entry.RuleID, entry.Path)
	}
	return summary.String()
}

File: rust/crates/slophammer-cli/src/baseline.rs
use crate::core::{Finding, Report};
use serde::{Deserialize, Serialize};
use std::collections::BTreeSet;
use std::path::Path;
use thiserror::Error;

pub const BASELINE_FILE: &str = "slophammer-baseline.json";

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub enum BaselineMode {
    #[default]
    Off,
    Check,
    Write,
}

#[derive(Debug, Error)]
pub enum BaselineError {
    #[error("baseline file {BASELINE_FILE} is missing")]
    Missing,
    #[error("baseline parse failed: {0}")]
    Parse(String),
    #[error("baseline version must be 1")]
    Version,
    #[error("baseline contains resolved findings; rewrite it: {0}")]
    Stale(String),
    #[error("baseline write would grow the baseline; fix the new findings instead: {0}")]
    Superset(String),
    #[error("baseline write failed: {0}")]
    Write(String),
}

#[derive(Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
struct BaselineFile {
    version: u32,
    findings: Vec<BaselineEntry>,
}

#[derive(Clone, Deserialize, Eq, Ord, PartialEq, PartialOrd, Serialize)]
#[serde(deny_unknown_fields)]
struct BaselineEntry {
    rule_id: String,
    path: String,
}

/// Applies a checked-in baseline to a report: matched findings are marked
/// baselined and stop affecting `ok`; stale entries are an error so the
/// ratchet can only shrink. Matching is on rule_id plus path, never message.
pub fn apply_check(root: &Path, report: &mut Report) -> Result<(), BaselineError> {
    let baseline = read_baseline(root)?;
    let mut unmatched = baseline.clone();
    for finding in &mut report.findings {
        if baseline.contains(&entry_of(finding)) {
            finding.baselined = Some(true);
            unmatched.remove(&entry_of(finding));
        }
    }
    if !unmatched.is_empty() {
        return Err(BaselineError::Stale(joined(&unmatched)));
    }
    report.ok = report
        .findings
        .iter()
        .all(|finding| finding.baselined == Some(true));
    Ok(())
}

/// Records current findings as the baseline. Refuses to write a superset of
/// an existing baseline and prints the added and removed entries, so debt is
/// recorded once, reviewed, and only ever reduced.
pub fn write(root: &Path, report: &Report) -> Result<String, BaselineError> {
    let current: BTreeSet<BaselineEntry> = report.findings.iter().map(entry_of).collect();
    let previous = read_baseline(root).unwrap_or_default();
    let added: BTreeSet<_> = current.difference(&previous).cloned().collect();
    let removed: BTreeSet<_> = previous.difference(&current).cloned().collect();
    if !previous.is_empty() && !added.is_empty() && removed.is_empty() {
        return Err(BaselineError::Superset(joined(&added)));
    }
    let file = BaselineFile {
        version: 1,
        findings: current.iter().cloned().collect(),
    };
    let serialized = serde_json::to_string_pretty(&file)
        .map_err(|error| BaselineError::Write(error.to_string()))?;
    std::fs::write(root.join(BASELINE_FILE), format!("{serialized}\n"))
        .map_err(|error| BaselineError::Write(error.to_string()))?;
    Ok(write_summary(current.len(), &added, &removed))
}

pub fn debt_line(report: &Report) -> String {
    let baselined = report
        .findings
        .iter()
        .filter(|finding| finding.baselined == Some(true))
        .count();
    let new = report.findings.len() - baselined;
    format!("{baselined} findings baselined; {new} new\n")
}

fn read_baseline(root: &Path) -> Result<BTreeSet<BaselineEntry>, BaselineError> {
    let content =
        std::fs::read_to_string(root.join(BASELINE_FILE)).map_err(|_| BaselineError::Missing)?;
    let file: BaselineFile =
        serde_json::from_str(&content).map_err(|error| BaselineError::Parse(error.to_string()))?;
    if file.version != 1 {
        return Err(BaselineError::Version);
    }
    Ok(file.findings.into_iter().collect())
}

fn entry_of(finding: &Finding) -> BaselineEntry {
    BaselineEntry {
        rule_id: finding.rule_id.clone(),
        path: finding.path.clone(),
    }
}

fn joined(entries: &BTreeSet<BaselineEntry>) -> String {
    entries
        .iter()
        .map(|entry| format!("{} at {}", entry.rule_id, entry.path))
        .collect::<Vec<_>>()
        .join(", ")
}

fn write_summary(
    total: usize,
    added: &BTreeSet<BaselineEntry>,
    removed: &BTreeSet<BaselineEntry>,
) -> String {
    let mut summary = format!("baseline written: {total} finding(s)\n");
    for entry in added {
        summary.push_str(&format!("added: {} at {}\n", entry.rule_id, entry.path));
    }
    for entry in removed {
        summary.push_str(&format!("removed: {} at {}\n", entry.rule_id, entry.path));
    }
    summary
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::Severity;
    use crate::report::new_report;

    fn finding(rule_id: &str, path: &str) -> Finding {
        Finding {
            rule_id: rule_id.to_owned(),
            severity: Severity::Error,
            path: path.to_owned(),
            message: "missing".to_owned(),
            baselined: None,
        }
    }

    fn write_baseline(root: &Path, findings: &[(&str, &str)]) {
        let entries = findings
            .iter()
            .map(|(rule_id, path)| BaselineEntry {
                rule_id: (*rule_id).to_owned(),
                path: (*path).to_owned(),
            })
            .collect();
        let file = BaselineFile {
            version: 1,
            findings: entries,
        };
        std::fs::write(
            root.join(BASELINE_FILE),
            serde_json::to_string(&file).expect("serialize"),
        )
        .expect("write baseline");
    }

    #[test]
    fn baselined_findings_stop_affecting_ok() {
        let dir = tempfile::tempdir().expect("tempdir");
        write_baseline(dir.path(), &[("repo.readme-required", "README.md")]);
        let mut report = new_report(vec![finding("repo.readme-required", "README.md")]);

        apply_check(dir.path(), &mut report).expect("apply baseline");

        assert!(report.ok);
        assert_eq!(report.findings[0].baselined, Some(true));
    }

    #[test]
    fn new_findings_keep_failing() {
        let dir = tempfile::tempdir().expect("tempdir");
        write_baseline(dir.path(), &[("repo.readme-required", "README.md")]);
        let mut report = new_report(vec![
            finding("repo.readme-required", "README.md"),
            finding("repo.agents-required", "AGENTS.md"),
        ]);

        apply_check(dir.path(), &mut report).expect("apply baseline");

        assert!(!report.ok);
        assert_eq!(debt_line(&report), "1 findings baselined; 1 new\n");
    }

    #[test]
    fn stale_entries_are_an_error() {
        let dir = tempfile::tempdir().expect("tempdir");
        write_baseline(dir.path(), &[("repo.readme-required", "README.md")]);
        let mut report = new_report(Vec::new());

        let error = apply_check(dir.path(), &mut report).expect_err("stale baseline");

        assert!(error.to_string().contains("resolved findings"));
    }

    #[test]
    fn write_refuses_supersets() {
        let dir = tempfile::tempdir().expect("tempdir");
        write_baseline(dir.path(), &[("repo.readme-required", "README.md")]);
        let report = new_report(vec![
            finding("repo.readme-required", "README.md"),
            finding("repo.agents-required", "AGENTS.md"),
        ]);

        let error = write(dir.path(), &report).expect_err("superset write");

        assert!(error.to_string().contains("grow the baseline"));
    }

    #[test]
    fn write_records_and_shrinks() {
        let dir = tempfile::tempdir().expect("tempdir");
        let first = new_report(vec![
            finding("repo.readme-required", "README.md"),
            finding("repo.agents-required", "AGENTS.md"),
        ]);
        let summary = write(dir.path(), &first).expect("initial write");
        assert!(summary.contains("baseline written: 2"));

        let second = new_report(vec![finding("repo.agents-required", "AGENTS.md")]);
        let summary = write(dir.path(), &second).expect("shrinking write");
        assert!(summary.contains("removed: repo.readme-required at README.md"));
    }
}
